rust/connlib/tunnel/src/client.rs

```diff
@@ -644,6 +641,21 @@ impl IpProvider {
             ipv6: Box::new(ipv6.subnets_with_prefix(128).map(|ip| ip.network_address())),
         }
     }
+
+    pub fn get_ip_for(&mut self, ip: &IpAddr) -> Option<IpAddr> {
+        let ip = match ip {
+            IpAddr::V4(_) => self.ipv4.next().map(Into::into),
+            IpAddr::V6(_) => self.ipv6.next().map(Into::into),
+        };
+
+        if ip.is_none() {
+            // TODO: we might want to make the iterator cyclic or another strategy to prevent ip exhaustion
+            // this might happen in ipv4 if tokens are too long lived.
+            tracing::error!("IP exhaustion: Please reset your client");
```

Comment:
Are `tracing::error` calls supposed to bubble up to the GUI?

Reply:
no but we are planning to completely get rid of `on_error` with #2867 this is just for users that look at the logs.